### PR TITLE
Prefix $path for filename for internal file cache

### PR DESCRIPTION
**Pull request by kesselb**
Follow-Up https://github.com/nextcloud/server/pull/11518

- `file1.txt`
- `folder/file1.txt`

If a file with the same name exists in a sub directory (prefix to be precise) `file1.txt` is overwritten (the array key not the file itself) by `folder/file1.txt`. `getContentLength` and `getLastModified` are called with `folder/file1.txt` and the internal cache does not hit.

### CODE REVIEW

File: apps/files_external/lib/Lib/Storage/AmazonS3.php [hunk 1]
```diff
@@ -408,7 +408,7 @@ public function filetype($path) {
 		}
 
 		try {
-			if ($this->headObject($path)) {
+			if (isset($this->filesCache[$path]) && $this->headObject($path)) {
```
Comment (icewind1991) [inline apps/files_external/lib/Lib/Storage/AmazonS3.php]: what is the reasoning here?

Comment (kesselb) [reply apps/files_external/lib/Lib/Storage/AmazonS3.php]: If `filesCache` has been populated by `opendir` we already know that path is a file or not without calling headObject. If you wonder about "only call headObject if $path isset in filescache" https://github.com/nextcloud/server/pull/14211/commits/0d667d18bb9d76e8ffad48b270545dde76fc156e

**Review by icewind1991 — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]
Code makes sense 👍

**Comment by MorrisJobke** [comment]
@kesselb @icewind1991 Backport to stable15?

**Pull request merged**